Lab01 - Hardware introduction (#34)
UPB-FILS-MA/upb-fils-ma.github.io: docs/lab/01.md

sidebar_position: 2
---

# 01 - Hardware Introduction

> We will use [Kicad](https://www.kicad.org/) Electronics Design Suite for the electronic schematics.

Reviewer: ```suggestion
We will use [KiCad](https://www.kicad.org/) Electronics Design Suite for the electronic schematics.
```

## Resources

> 1. [Getting Started in Kicad](https://docs.kicad.org/7.0/en/getting_started_in_kicad/getting_started_in_kicad.html) 

Reviewer: ```suggestion
1. [Getting Started in KiCad](https://docs.kicad.org/7.0/en/getting_started_in_kicad/getting_started_in_kicad.html) 
```

> 2. **Charles Platt**, *Encyclopedia of Electronic Components*, Volumes 1, 2, 3 

Reviewer: Ping here

sidebar_position: 2
---

# 01 - Hardware Introduction

We will use [KiCad](https://www.kicad.org/) Electronics Design Suite for the electronic schematics.

## Resources
1. [Getting Started in KiCad](https://docs.kicad.org/7.0/en/getting_started_in_kicad/getting_started_in_kicad.html) 
2. **Charles Platt**, *Encyclopedia of Electronic Components*, Volumes 1, 2, 3 

## Basic electronics

### Definitions

#### Electronic voltage

Electronic voltage represents the potential difference between two points in a circuit and it's proportional to the energy required to move an electric charge between the two points.

$$

> U = \frac{L}{Q};

Reviewer: ```suggestion
U = \frac{L}{Q}
```

$$

### Ohm's law

The intensity(I) of the electric current passing through a resistor is directly proportional to the voltage(U) applied to the resistor and inversely proportional to its resistance value(R).

<div align="center">
![Ohm's Law](images/ohm.png)
</div>

$$
I = \frac{U}{R};
$$

I = intensity of electric current(A)   
U = electrical voltage(V)   
R = circuit's resistance(&ohm;)

### Kirchhoff's First Law

The sum of currents flowing into a junction is always equal to the sum of currents flowing out of the same junction.

$$
\sum_{k=1}^{n} I_k = 0
$$

<div align="center">
![Kirchhoff's First Law](images/kirchhoff_1.png)
</div>

$$
I_1 + I_2 = I_3 + I_4 + I_5 \\
I_1 + I_2 - I_3 - I_4 - I_5 = 0
$$

### Kirchhoff's Second Law

Around any closed loop in a circuit, the directed sum of potential differences across components is zero.

$$
\sum_{k=1}^{n} U_k = 0
$$

<div align="center">
![Kirchhoff's Second Law](images/kirchhoff_2.png)
</div>

$$
-U_A - U_B - U_C + U_D + U_E = 0
$$

### Voltage divider

There are many types of voltage dividers, being named according to the type of component elements: resistive divider, capacitive divider, compensated divider, etc...   

> The **resistive voltage divider** is obtained by applying a voltage `E` to a group of resistors in series and thus a fraction of the voltage applied to one of the resistors of the group can be obtained.

Reviewer: ```suggestion
The **resistive voltage divider** is obtained by applying a voltage $$E$$ to a group of resistors in series and thus a fraction of the voltage applied to one of the resistors of the group can be obtained.
```

<div align="center">
![Voltage divider on GND](images/voltage_divider_gnd.png)
</div>

$$
V_{out} = V_{in} * \frac{R_{2}}{R_{1} + R_{2}};
$$

:::warning

> Voltage divider can be considered a power source only if it operates in the empty state. It is **not recommended** to use a voltage divider to supply a circuit. It can be used to provide reference points.

Reviewer: I would explain why => the internal resistance is high and energy is lost due to heating.

:::

#### What happens if we use the voltage divider to supply a circuit.

> We consider a voltage divider that provides `3V3` from a `5V` power source `E` and a load resistance, `Rs`, representing the current consumption of a sensor or a circuit that needs to be supplied at `3V3`.

Reviewer: ```suggestion
We consider a voltage divider that provides $$3V3$$ from a $$5V$$ power source $$E$$ and a load resistance, $$R_s$$, representing the current consumption of a sensor or a circuit that needs to be supplied at $$3V3$$.
```

<div align="center">
![Voltage divider with load](images/voltage_divider_rs.png)
</div>

> From Kirchhoff's First Law, the current through `R1` must be equal to the sum of the current through `R2` and `Rs`. 

Reviewer: ```suggestion
From Kirchhoff's First Law, the current through $$R_1$$ must be equal to the sum of the current through $$R_2$$ and $$R_s$$. 
```

$$
I_{R_1} = I_{R_2} + I_{R_S} 

U = U_{E} * \frac{R_{2} || R_{S}}{R_{1} + R_{2} || R_{S}};
$$

> The output voltage depends on the current intensity through `Rs`, on the current consumption of the circuit that needs to be supplied at `3V3`. This is not a viable power supply solution.

Reviewer: ```suggestion
The output voltage depends on the current intensity through $$R_s$$, on the current consumption of the circuit that needs to be supplied at $$3V3$$. This is not a viable power supply solution.
```

:::note

> Beside the instability of the voltage divider with a load, the power rating for the `R1` must be suitable.

Reviewer: ```suggestion
Beside the instability of the voltage divider with a load, the power rating for the $$R_1$$ must be suitable.
```

$$
P_{R_1} = U_{R_1} * I_{R_1}
$$

> The power dissipation on the resistor is directly proportional with the current through the resistor. Resistor are fabricated with predefined power ratings, most common `1/4W`, `1/2W`, `1W`. 

Reviewer: ```suggestion
The power dissipation on the resistor is directly proportional with the current through the resistor. Resistor are fabricated with predefined power ratings, most common $$\frac{1}{4}W$$, `1/2W`, `1W`. 
```

Not sure about this though.

Author: Not sure about the statement, or about the style?

:::

:::info
In case we want to use the voltage divider between two chosen voltage values, we can use the generalized formula:

$$
V_{out} = ( V_{1} - V_{2} ) * \frac{R_{2}}{R_{1} + R_{2}};
$$

<div align="center">
![Voltage divider](images/voltage_divider.png)
</div>
:::

> 

Reviewer: ```suggestion

## Electronic Components

```

#### Calculation of current limiting resistor

To use an LED for the purpose of indicating the status of a pin (rather said to indicate the presence of voltage), the current through the LED must be limited. This can be done most simply by stringing a resistor with the LED.   

An LED is designed to operate at a nominal current (ex: 10mA). The voltage drop at this current across low power indicator LEDs is given by the chemistry of the LED (this also gives the color of the LED). In the lab, since we are using such a low current LED, we can power it directly from the logic pins of the MCU.

<div align="center">
![LED](images/led.png)
</div>

$$
R_{1} = \frac{(U_{pin} - U_{led})}{I_{led}}
$$

#### Example:

If the MCU has a pin voltage of 3.3V, also noted as 3V3, to light up an LED with a nominal current of 10mA and a voltage drop of 2V we need 
a resistance of 130 &ohm;.

:::note
We can use a resistor with a higher resistance value. The nominal current will light up the LED at it's maximum brightness. For status LEDs we can pick a resistance even 10 times bigger and the LED will light up slightly. 
:::

:::danger

> If there is no resistor in the circuit, the resistance will be almost `0`, the current will tend to `infinity`, meaning a short circuit. This will absolutely burn the LED and make it unusable, but it can also burn the MCU. Most MCUs have short circuit protection, but is safer to not rely on that.  

Reviewer: ```suggestion
If there is no resistor in the circuit, the resistance will be almost 0 &ohm;, the current will tend to `infinity`, meaning a short circuit. This will absolutely burn the LED and make it unusable, but it can also burn the MCU. Most MCUs have short circuit protection, but is safer to not rely on that.  
```

:::

### Buttons

The simplest way for the user to interact with a MCU is through the use of buttons.

There are various ways to connect a button to the MCU, but these are the most used versions:

#### a)
<div align="center">
![Floating button](images/floating_btn.png)
</div>

#### b)
<div align="center">
![Pulled up button](images/pullup_btn.png)
</div>

> a) It shows a button connected to the MCU pin. When the button is pressed, the MCU input pin will be connected to GND, so it will be in the logic "0" state. This way of binding is **incorrect** because when the button is not pressed, the input is in **an undefined state** (as if left in the air), not being connected to either GND or Vcc! This state is called the **increased impedance state**. In practice, if we now read the value of the pin, it will produce a result of 1 or 0 depending on the environmental conditions. For example, if we bring our finger closer to that input, the reading will be 1, and if we move our finger away, the reading will be 0. 

Reviewer: Tip with danger or warning and title Incorrect unde the image?

> b) It shows the correct way to connect the button, using a **pull-up resistor** between the input pin and Vcc. This resistance has the role of bringing the input to the logic "1" state when the button is free by "raising" the line potential to Vcc. Alternatively, a **pull-down resistor** (connected to GND) can be used, in which case the input is held in the logic "0" state while the button is not pressed.

Reviewer: Tip with green and title Correct under the image?

sidebar_position: 2
---

# 01 - Hardware Introduction

We will use [KiCad](https://www.kicad.org/) Electronics Design Suite for the electronic schematics.

## Resources
1. [Getting Started in KiCad](https://docs.kicad.org/7.0/en/getting_started_in_kicad/getting_started_in_kicad.html) 
2. **Charles Platt**, *Encyclopedia of Electronic Components*, Volumes 1, 2, 3 

## Basic electronics

### Definitions

#### Electronic voltage

Electronic voltage represents the potential difference between two points in a circuit and it's proportional to the energy required to move an electric charge between the two points.

$$

> U = \frac{L}{Q}

Reviewer 2: In English, mechanical work is usually denoted by W, and voltage by V.

<div align="center">
![LED](images/led.png)
</div>

$$
R_{1} = \frac{(U_{pin} - U_{led})}{I_{led}}
$$

#### Example:

If the MCU has a pin voltage of 3.3V, also noted as 3V3, to light up an LED with a nominal current of 10mA and a voltage drop of 2V we need 
a resistance of 130 &ohm;.

:::note
We can use a resistor with a higher resistance value. The nominal current will light up the LED at it's maximum brightness. For status LEDs we can pick a resistance even 10 times bigger and the LED will light up slightly. 
:::

:::danger
If there is no resistor in the circuit, the resistance will be almost 0 &ohm;, the current will tend to `infinity`, meaning a short circuit. This will absolutely burn the LED and make it unusable, but it can also burn the MCU. Most MCUs have short circuit protection, but is safer to not rely on that.  
:::

### Buttons

The simplest way for the user to interact with a MCU is through the use of buttons.

There are various ways to connect a button to the MCU, but these are the most used versions:

#### a)
<div align="center">
![Floating button](images/floating_btn.png)   
Incorrect
</div>

> 

Reviewer: I suggested putting this into a warning tip.

<div align="center">
![Footprint error ](images/footprint_error.png)   
Footprint error 
</div>

Go back to the Schematic and assign footprints for resistor $$R1$$ and LED $$D1$$ like in the setup [tutorial](../tutorials/kicad.md). You can pick any footprint, or just paste these on footprint field in symbol properties.
- resistor: Resistor_THT:R_Axial_DIN0204_L3.6mm_D1.6mm_P2.54mm_Vertical
- LED: LED_THT:LED_D3.0mm

Go back to the PCB editor and try to update again.

<div align="center">
![Update PCB](images/update_pcb_2.png)   
Update PCB
</div>

Connect the components using the `Route tracks` button on the right or the shortcut `X`. The blue lines will guide you to connect the right pins.

<div align="center">
![Route tracks](images/route_tracks.png)   
Route tracks
</div>

KiCad offers a 3D Viewer to have a preview of how the PCB will look. Click on `View` -> `3D Viewer`

<div align="center">
![3D View](images/3D_view.png)   
3D View
</div>

:::info
For a better understanding, please read [Getting Started in KiCad](https://docs.kicad.org/7.0/en/getting_started_in_kicad/getting_started_in_kicad.html) tutorial.
:::

## Exercises

> 1. Complete the given exercise sheet.(**3p**)

Reviewer: Add a download link here to the pdf, uou can add it in assets.

Author: I want to make different versions, print them and give it to the students

2. The microcontroller is already running a predefined program. Assemble a circuit surrounding the microcontroller in order to perform the next tasks:(**3p**)
- A status LED is ON as long as the microcontroller is supplied. The pin that determines the status of the LED is TBD.
- A button is used to toggle the LED integrated on the development board. The pin reading the state of the button is TBD.
- On the screen is displayed an analog value read by the microcontroller on pin TBD. Create a circuit in order to display the value TBD(can be customized) on the screen.

:::warning

> Do not supply the microcontroller before the lab teacher checks the circuit.

Reviewer: ```suggestion
Do not supply power to the microcontroller before the lab teacher checks the circuit.
```

$$

:::info

Electronic voltage is always measured between two points in a circuit. In general, voltages are measured against a reference point called circuit mass **(GND)**. Circuit mass is a convention and represents the point which potential is consider null. 

:::

#### Electric resistance

Electrical resistance is a physical quantity that expresses the property of a material to oppose the passage of electric current. The SI[^si] unit of resistance is the **Ohm**, noted &ohm;.

$$
[R]_{SI} = \Omega(Ohm)
$$

#### Intensity of electric current

The intensity of the electric current, also called the electric current, is a scalar physical quantity equal to the variation of the electric charge that crosses the section of a conductor in the unit of time. The unit of measurement in SI[^si] is the **Ampere**(A).

$$
[I]_{SI} = A(Ampere)
$$

### Ohm's law

The intensity(I) of the electric current passing through a resistor is directly proportional to the voltage(U) applied to the resistor and inversely proportional to its resistance value(R).

<div align="center">
![Ohm's Law](images/ohm.png)
</div>

$$
I = \frac{U}{R};
$$

I = intensity of electric current(A)   

> U = electrical voltage(V)   

Reviewer: ```suggestion
V = electrical voltage(V)   
```

R = circuit's resistance(&ohm;)

### Kirchhoff's First Law

The sum of currents flowing into a junction is always equal to the sum of currents flowing out of the same junction.

$$
\sum_{k=1}^{n} I_k = 0
$$

<div align="center">
![Kirchhoff's First Law](images/kirchhoff_1.png)
</div>

$$
I_1 + I_2 = I_3 + I_4 + I_5 \\
I_1 + I_2 - I_3 - I_4 - I_5 = 0
$$

### Kirchhoff's Second Law

Around any closed loop in a circuit, the directed sum of potential differences across components is zero.

$$

> \sum_{k=1}^{n} U_k = 0

Reviewer: ```suggestion
\sum_{k=1}^{n} V_k = 0
```

$$

<div align="center">
![Kirchhoff's Second Law](images/kirchhoff_2.png)
</div>

$$

> -U_A - U_B - U_C + U_D + U_E = 0

Reviewer: ```suggestion
-V_A - V_B - V_C + V_D + V_E = 0
```

The **resistive voltage divider** is obtained by applying a voltage $$E$$ to a group of resistors in series and thus a fraction of the voltage applied to one of the resistors of the group can be obtained.

<div align="center">
![Voltage divider on GND](images/voltage_divider_gnd.png)
</div>

$$
V_{out} = V_{in} * \frac{R_{2}}{R_{1} + R_{2}};
$$

:::warning
Voltage divider can be considered a power source only if it operates in the empty state. It is **not recommended** to use a voltage divider to supply a circuit as the internal resistance is high and energy is lost due to heating. It can be used to provide reference points.
:::
#### What happens if we use the voltage divider to supply a circuit.

We consider a voltage divider that provides $$3V3$$ from a $$5V$$ power source $$E$$ and a load resistance, $$R_s$$, representing the current consumption of a sensor or a circuit that needs to be supplied at $$3V3$$.

<div align="center">
![Voltage divider with load](images/voltage_divider_rs.png)
</div>

From Kirchhoff's First Law, the current through $$R_1$$ must be equal to the sum of the current through $$R_2$$ and $$R_s$$. 

$$
I_{R_1} = I_{R_2} + I_{R_S} 

> U = U_{E} * \frac{R_2 || R_S}{R_1 + R_2 || R_S};

Reviewer: ```suggestion
V = V_{E} * \frac{R_2 || R_S}{R_1 + R_2 || R_S};
```

:::

### LEDs

LEDs - Light Emitting Diode - also called electroluminescent diodes - emit light when they are directly polarized. Not to be confused with light bulbs as they have radically different methods of operation.   

LEDs can be used as indicator lights (often used in various appliances to signal that the appliance is on and doing something), or for illumination, in which case power LEDs are used. In the lab, LEDs are used to indicate the status of a pin.

#### Calculation of current limiting resistor

To use an LED for the purpose of indicating the status of a pin (rather said to indicate the presence of voltage), the current through the LED must be limited. This can be done most simply by stringing a resistor with the LED.   

An LED is designed to operate at a nominal current (ex: 10mA). The voltage drop at this current across low power indicator LEDs is given by the chemistry of the LED (this also gives the color of the LED). In the lab, since we are using such a low current LED, we can power it directly from the logic pins of the MCU.

<div align="center">
![LED](images/led.png)
</div>

$$

> R_{1} = \frac{(U_{pin} - U_{led})}{I_{led}}

Reviewer: ```suggestion
R_{1} = \frac{(U_{pin} - V_{led})}{I_{led}}
```

:::

### Breadboard and jumper wires

#### Breadboard

A breadboard is a rectangular board with a grid of holes that allows you to create temporary electronic circuits without soldering. The board typically has metal strips underneath the surface, connecting the holes in certain patterns. These patterns follow a standard layout, facilitating circuit building. Breadboards are reusable and provide a convenient way to prototype circuits quickly and make changes easily by rearranging components.

<div align="center">
![Breadboard](images/breadboard.png)   
![Breadboard connection](images/breadboard_connections.png)   
Breadboard connection   
![Small breadboard connection](images/small_breadboard.png)   
Small breadboard connection
</div>

#### Jumper wires

Jumper wires are flexible wires with connectors at each end, typically male connectors (pins) or female connectors (sockets). They are used to create electrical connections on a breadboard by plugging one end into a hole on the breadboard and the other end into another hole, forming a connection between the two points.

<div align="center">
![Jumper wires](images/jumper_wires.png)   
Jumper wires
</div>

> :::info

Reviewer: ```suggestion
:::note
```

> For a better understanding, please read the chapter 1 *Encyclopedia of Electronic Components*, Volume 1

Reviewer: ```suggestion
For a better understanding, please read the chapter 1 from *Encyclopedia of Electronic Components*, Volume 1
```
